src/persist/src/indexed/mod.rs

```diff
@@ -1308,6 +1503,253 @@ mod tests {
+
+    /// Test the communication protocol between [Indexed] and [Maintainer]
+    /// specifically for trace compaction requests.
+    #[test]
+    fn trace_compaction() -> Result<(), Error> {
+        let mut unreliable = UnreliableHandle::default();
+        let (mut i, maintainer, id) = trace_compaction_setup(unreliable.clone())?;
+
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        let request = reqs[0].clone();
+
+        // Once a request is in flight, don't send requests for the same stream.
+        let reqs = i.step()?;
+        assert_eq!(reqs, vec![]);
+
+        // Handle receiving an error MaintenanceRes.
+        let error_response =
+            MaintenanceRes::CompactTrace((id, Err(Error::from("test compaction error"))));
+        i.handle_maintenance_res(error_response.clone());
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // After receiving an error, retry the original request.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        assert_eq!(reqs[0], request);
+
+        // Handle receiving a valid MaintenanceRes.
+        let response = get_maintenance_response(&maintainer, request.clone());
+
+        i.handle_maintenance_res(response);
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // After receiving a valid response, issue a new compaction request.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        let request2 = reqs[0].clone();
+        assert!(request2 != request);
+
+        // Construct a new valid, response.
+        let response2 = get_maintenance_response(&maintainer, request2.clone());
+
+        // Send back an error response. After this there are no requests in flight.
+        i.handle_maintenance_res(error_response);
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // Retry request after error response.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        assert_eq!(reqs[0], request2);
+
+        // Handle failed write to metadata after receiving a response.
+        i.handle_maintenance_res(response2.clone());
+        unreliable.make_unavailable();
+        assert_eq!(i.drain_pending(), Err(Error::from("unavailable: blob set")));
+        unreliable.make_available();
+
+        // Retry request after failed meta write.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        assert_eq!(reqs[0], request2);
+
+        // Handle receiving a response after the trace is deleted.
+        block_on(|res| i.destroy("0", res))?;
+
+        // Handle receiving an valid MaintenanceRes for a stream that has been deleted.
+        i.handle_maintenance_res(response2);
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // Don't produce any more requests because the stream is deleted.
+        let reqs = i.step()?;
+        assert_eq!(reqs, vec![]);
+
+        Ok(())
+    }
+
+    /// This test checks that Indexed correctly handles receiving an invalid MainenanceRes
```

Comment:
nit: s/invalid/unexpected/

Reply:
the wording was intentional to try to frame unexpected as "got a response when none was in flight" vs invalid "got a response that didn't match the in flight request" mostly because i needed two different names for the two test functions

Reply:
gotcha. still let's name this something different. I elaborate on this in another comment, but "invalid" means something specific to me that doesn't apply here

Reply:
fixed here and elsewhere.